autovalet_lane_detection/scripts/autovalet_lane_detector.py

```diff
@@ -0,0 +1,186 @@
+        edges = cv2.Canny(roi, 100, 200, apertureSize=3)
+        lines = cv2.HoughLinesP(edges, 1, np.pi/180, 30, np.array([]), minLineLength, maxLineGap)
+
+        # possible scenarios: 1. new line 2. tracking
+        max_len, scenario, X1, Y1, X2, Y2 = 0, 0, 0, 0, 0, 0
+        if lines is not None: # 1. new line
+            for line in lines:
+                x1, y1, x2, y2 = line[0]
+                line_len = np.sqrt((x2-x1)**2 + (y2-y1)**2)
+                if max_len < line_len:
+                    max_len = line_len
+                    X1, Y1, X2, Y2 = x1, y1, x2, y2
+            scenario = 1
+            bb = (X1, Y1, np.abs(X2-X1), np.abs(Y2-Y1))
+            self.tracker = cv2.TrackerKCF_create()
+            self.tracker.init(rgb, bb)
+            self.lose_track_count = 0
+        elif self.tracker is not None:
+            success, bb = self.tracker.update(rgb)
+            x, y, w, h = [int(v) for v in bb]
+            if success: # 2. tracking
+                scenario = 2
+                X1, Y1, X2, Y2 = x, y, x+w, y-h
+                Y1 = min(Y1, img.shape[0] - 1)
+
+
+        # coordinates and slope
+        if scenario == 0:
+            return np.array([])
+        if X1 == X2: # vertical line
+            coordinates = [[X1, i] for i in range(Y2, Y1+1)]
+        else:
+            slope = 1.0 * (Y2 - Y1) / (X1 - X2) # assuming slope being positive, might have bug
+            if slope > 1:
+                coordinates = [[X1+i, int(Y1-i*slope)] for i in range(X2-X1)]
+            else:
+                coordinates = [[int(X1+i/slope), Y1-i] for i in range(Y1-Y2)]
+
+        return np.array(coordinates)
+
+    def line2cloud(self, depth, coordinates):
+        # ref: Subbu's code
+        u, v = coordinates[:, 0], coordinates[:, 1]
+        x    = (u - self.cx) / self.fx
+        y    = (v - self.cy) / self.fy
+
+        z = depth[v, u] / 1000.0;
+        x = np.multiply(x, z)
+        y = np.multiply(y, z)-2  #: tmp fix accounted for map being gen w.r.t base_link
+
+        cloud = np.hstack((x.reshape(-1, 1), y.reshape(-1, 1), z.reshape(-1, 1)))
+
+        return cloud
+
+    def interpolateAnotherLine(self, center_line_cloud):
+        # extend each (x,y) along a normal vector of the center line
+        # toward right, for lane_width meters
+        lane_width = 3.5 # meters, US standard
+
+        center_x = center_line_cloud[:, 0]
+        center_y = center_line_cloud[:, 1]
+        center_z = center_line_cloud[:, 2]
+
+        # calculate the mean of the points
+        center_mean = center_line_cloud.mean(axis=0)
+
+        # Do an SVD on the mean-centered data.
+        u, d, vh = np.linalg.svd(center_line_cloud - center_mean)
+        orth_vec = np.sign(vh[0][2]) * np.array([vh[0][2], -vh[0][0]])
+        #orth_vec = np.sign(vh[-1][0]) * np.array([vh[-1][0], vh[-1][2]])
+        norm_vec = orth_vec / np.linalg.norm(orth_vec)
+
+        # extend the points with lane_width
```

Comment:
what frame's x,y,z? mention that here

Reply:
put some comments right at the start of the function.